crates/next-core/src/manifest.rs

```diff
@@ -0,0 +1,103 @@
+use anyhow::Result;
+use indexmap::IndexSet;
+use mime::APPLICATION_JSON;
+use turbo_tasks::primitives::StringsVc;
+use turbo_tasks_fs::File;
+use turbopack_core::asset::AssetContentVc;
+use turbopack_dev_server::source::{
+    combined::CombinedContentSourceVc, conditional::ConditionalContentSourceVc, ContentSource,
+    ContentSourceContent, ContentSourceData, ContentSourceResultVc, ContentSourceVc,
+};
+use turbopack_node::{
+    node_api_source::NodeApiContentSourceVc, node_rendered_source::NodeRenderContentSourceVc,
+};
+
+#[turbo_tasks::value(shared)]
+pub struct DevManifestContentSource {
+    pub page_roots: Vec<ContentSourceVc>,
+}
+
+#[turbo_tasks::value_impl]
+impl DevManifestContentSourceVc {
+    #[turbo_tasks::function]
+    async fn find_routes(self) -> Result<StringsVc> {
+        let this = &*self.await?;
+        let mut queue = this.page_roots.clone();
+        let mut routes = IndexSet::new();
+
+        while let Some(content_source) = queue.pop() {
+            if let Some(combined_source) =
+                CombinedContentSourceVc::resolve_from(content_source).await?
+            {
+                queue.extend(combined_source.await?.sources.iter().copied());
+
+                continue;
+            }
+
+            if let Some(conditional_source) =
+                ConditionalContentSourceVc::resolve_from(content_source).await?
+            {
+                let conditional_source = &*conditional_source.await?;
+                queue.push(conditional_source.activator);
+                queue.push(conditional_source.action);
+
+                continue;
+            }
+
+            if let Some(api_source) = NodeApiContentSourceVc::resolve_from(content_source).await? {
+                routes.insert(format!("/{}", api_source.get_pathname().await?));
+
+                continue;
+            }
+
+            if let Some(page_source) =
+                NodeRenderContentSourceVc::resolve_from(content_source).await?
+            {
+                routes.insert(format!("/{}", page_source.get_pathname().await?));
+
+                continue;
+            }
+
+            // ignore anything else
+        }
```

Comment:
Could we instead add two (optional?) methods on `ContentSourceVc`: `sources(self)` and `pathname(self)` Otherwise, when we add more content sources in the future, we might  forget to update this method.

Reply:
`sources` sounds like a good idea

`pathname` is kinda specific to the node sources right now and a lot of sources respond to multiple pathnames, so I'm not sure what the best way of doing that would be